src/NuGet.Core/NuGet.Packaging/Signing/Signatures/Signature.cs

```diff
@@ -31,6 +31,16 @@ public class Signature
         /// </summary>
         public SignatureContent SignatureContent { get; }
 
+        /// <summary>
+        /// Signature timestamp.
+        /// </summary>
+        public DateTimeOffset? Timestamp { get; }
```

Comment:
This is not fully correct. The timestamp can have a range of validaity based on the accuracy in the response. You can read more about it at Section 2.4.2 in [rfc3161](https://tools.ietf.org/html/rfc3161). The options - 

1.  The `TimestampSignerInfo` field has the relevant information that can be obtained, as needed.
2.  Add a field to remember the accuracy.

Reply:
Okay, so just extract the accuracy as well and store it?

Reply:
Should I return `AccuracyInMicroseconds` or convert that to a `TimeSpan`?

Reply:
I thought a bit about this and I think we should have the following - 

1. Enclose the timestamp in a class defined within `Signature` - 

```
public sealed class Timestamp 
{
    public DateTimeOffset UpperLimit { get; }
    public DateTimeOffset LowerLimit { get; }
    public SignerInfo TimestampSignerInfo { get; }
}
```

2. Change `GetTimestamp` method to return `Timestamp` type.

3. Add corresponding tests in a new file called `SignatureTests` to this [project](https://github.com/NuGet/NuGet.Client/tree/dev/test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests). The goal of the tests should be - Given a signed and timestamped package, [PackageArchiveReader.GetSignaturesAsync](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs#L230)
 should return a signature for which `Signature.Timestamp` is correctly populated. [Here](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/SignerTests.cs#L37) is an example of a test that generates a signed package. You will have to change `SignedArchiveTestUtility.CreateSignedPackageAsync` call to `SignedArchiveTestUtility.CreateSignedAndTimeStampedPackageAsync`.

Reply:
Should this be a nested type, `Signature.Timestamp` or a type alongside Signature?

Reply:
The type should not be nested.

Also, a tweak to Ankit's feedback:  the `TimestampSignerInfo` property should just be named `SignerInfo` since the property is already encapsulated by a class named `Timestamp`.